internal/hud/server/websocket_reader.go

```diff
@@ -0,0 +1,194 @@
+	handler      ViewHandler
+}
+
+func ProvideWebsockerReader() *WebsocketReader {
+	return &WebsocketReader{
+		// TODO(maia): pass this URL instead of hardcoding / wire this
+		url:          url.URL{Scheme: "ws", Host: "localhost:10350", Path: "/ws/view"},
+		handler:      NewLogStreamer(),
+		marshaller:   jsonpb.Marshaler{OrigName: false, EmitDefaults: true},
+		unmarshaller: jsonpb.Unmarshaler{},
+	}
+}
+
+type ViewHandler interface {
+	Handle(v proto_webview.View) error
+}
+
+type LogStreamer struct {
+	logstore   *logstore.LogStore
+	printer    *hud.IncrementalPrinter
+	checkpoint logstore.Checkpoint
+}
+
+func NewLogStreamer() *LogStreamer {
+	// TODO(maia): wire this (/ maybe this isn't the thing that needs to be wired, but
+	//   should be created after we have a conn to pass it?)
+	printer := hud.NewIncrementalPrinter(hud.Stdout(colorable.NewColorableStdout()))
+	return &LogStreamer{
+		logstore: logstore.NewLogStore(),
+		printer:  printer,
+	}
+}
+func (ls *LogStreamer) Handle(v proto_webview.View) error {
+	fromCheckpoint := logstore.Checkpoint(v.LogList.FromCheckpoint)
+	toCheckpoint := logstore.Checkpoint(v.LogList.ToCheckpoint)
+
+	if fromCheckpoint == -1 {
+		// Server has no new logs to send
+		return nil
+	}
+
+	segments := v.LogList.Segments
+	if fromCheckpoint < ls.checkpoint {
+		// The server is re-sending some logs we already have, so slice them off.
+		deleteCount := ls.checkpoint - fromCheckpoint
+		segments = segments[deleteCount:]
+	}
+
+	// TODO(maia): filter for the resources that we care about (`tilt logs resourceA resourceC`)
+	//   --> and if there's only one resource, don't prefix logs with resource name?
+	for _, seg := range segments {
+		// TODO(maia): secrets???
+		ls.logstore.Append(webview.LogSegmentToEvent(seg, v.LogList.Spans), model.SecretSet{})
+	}
+
+	ls.printer.Print(ls.logstore.ContinuingLines(ls.checkpoint))
+
+	if toCheckpoint > ls.checkpoint {
+		ls.checkpoint = toCheckpoint
+	}
+
+	return nil
+}
+
+func (wsr *WebsocketReader) Listen(ctx context.Context) error {
+	logger.Get(ctx).Debugf("connecting to %s", wsr.url.String())
+
+	var err error
+	wsr.conn, _, err = websocket.DefaultDialer.Dial(wsr.url.String(), nil)
+	if err != nil {
+		return errors.Wrapf(err, "dialing websocket %s", wsr.url.String())
+	}
+	defer wsr.conn.Close()
+
+	done := make(chan struct{})
+	go func() {
+		defer close(done)
+		for {
+			messageType, reader, err := wsr.conn.NextReader()
+			if err != nil {
+				// uh do i need to do anything with this error? or does it just mean that the socket has closed?
+				return
+			}
+
+			if messageType == websocket.TextMessage {
+				err = wsr.handleTextMessage(ctx, reader)
+				if err != nil {
+					// will I want this to be an Info sometimes??
+					logger.Get(ctx).Verbosef("Error handling websocket message: %v", err)
+				}
+			}
+		}
+	}()
+
+	for {
+		select {
+		case <-done:
+			return nil
+		case <-ctx.Done():
+			err := ctx.Err()
+			if err != context.Canceled {
+				return err
```

Comment:
@nicks i still don't know whether I should try to call wsr.conn.Close() here?

Reply:
honestly, your guess is as good as mine

Reply:
😬 👍